nginx-controller/controller/controller.go

```diff
@@ -1067,3 +1195,124 @@ func (lbc *LoadBalancerController) ValidateSecret(secret *api_v1.Secret) error {
 
 	return fmt.Errorf("Secret is not a TLS or JWK secret")
 }
+
+// getMinionsForHost returns a list of all minion ingress resources for a given master
+func (lbc *LoadBalancerController) getMinionsForMaster(master *nginx.IngressEx) ([]*nginx.IngressEx, error) {
+	ings, err := lbc.ingLister.List()
+	if err != nil {
+		return []*nginx.IngressEx{}, err
+	}
+
+	var minions []*nginx.IngressEx
+	for i, _ := range ings.Items {
+		if !lbc.isNginxIngress(&ings.Items[i]) {
+			continue
+		}
+		if !isMinion(&ings.Items[i]) {
+			continue
+		}
+		if ings.Items[i].Spec.Rules[0].Host != master.Ingress.Spec.Rules[0].Host {
+			continue
+		}
+		if len(ings.Items[i].Spec.Rules) != 1 {
+			glog.Errorf("Ingress Resource %v/%v with the 'nginx.org/mergible-ingress-type' annotation must contain only one host", ings.Items[i].Namespace, ings.Items[i].Name)
+			continue
+		}
+		if ings.Items[i].Spec.Rules[0].HTTP == nil {
+			glog.Errorf("Ingress Resource %v/%v with the 'nginx.org/mergible-ingress-type' annotation set to 'minion' must contain a Path", ings.Items[i].Namespace, ings.Items[i].Name)
+			continue
+		}
+		ingEx, err := lbc.createIngress(&ings.Items[i])
+		if err != nil {
+			glog.Errorf("Error creating ingress resource %v/%v: %v", ingEx.Ingress.Namespace, ingEx.Ingress.Name, err)
+			continue
+		}
+		if len(ingEx.TLSSecrets) > 0 || ingEx.JWTKey != nil {
+			glog.Errorf("Ingress Resource %v/%v with the 'nginx.org/mergible-ingress-type' annotation set to 'minion' cannot contain TLSSecrets or JWTKeys", ingEx.Ingress.Namespace, ingEx.Ingress.Name)
+			continue
+		}
+		minions = append(minions, ingEx)
+	}
+	return minions, nil
+}
+
+// findMasterForHost returns a master for a given minion
+func (lbc *LoadBalancerController) findMasterForMinion(minion *extensions.Ingress) (*extensions.Ingress, error) {
+	ings, err := lbc.ingLister.List()
+	if err != nil {
+		return &extensions.Ingress{}, err
+	}
+
+	for i, _ := range ings.Items {
+		if !lbc.isNginxIngress(&ings.Items[i]) {
+			continue
+		}
+		if !lbc.cnf.HasIngress(&ings.Items[i]) {
+			continue
+		}
+		if !isMaster(&ings.Items[i]) {
+			continue
+		}
+		if ings.Items[i].Spec.Rules[0].Host != minion.Spec.Rules[0].Host {
+			continue
+		}
+		return &ings.Items[i], nil
+	}
+
+	err = fmt.Errorf("Could not find a Master for Minion: '%v/%v'", minion.Namespace, minion.Name)
+	return &extensions.Ingress{}, err
+}
+
+func (lbc *LoadBalancerController) createMergableIngresses(master *extensions.Ingress) (*nginx.MergeableIngresses, error) {
+	mergeableIngresses := nginx.MergeableIngresses{}
+
+	if len(master.Spec.Rules) != 1 {
+		err := fmt.Errorf("Ingress Resource %v/%v with the 'nginx.org/mergible-ingress-type' annotation must contain only one host", master.Namespace, master.Name)
+		return &mergeableIngresses, err
+	}
+
+	var empty extensions.HTTPIngressRuleValue
```

Comment:
I'll refactor this in another PR, but if:

`if master.Spec.Rules[0].HTTP != &empty` is not checked after `master.Spec.Rules[0].HTTP != nil` then there will be a null pointer. I can combined the other 2.

Reply:
got it